lark/parsers/lalr_analysis.py

```diff
@@ -279,9 +282,97 @@ def compute_lalr1_states(self):
         else:
             self.parse_table = IntParseTable.from_ParseTable(_parse_table)
 
+    # Generation of conflict resolution hints
+    def __build_reduce_reduce_error(self, la, rules):
+        """Build a GrammarError with conflict resolution hints."""
+        msg = "Reduce/Reduce collision in %s between the following rules:" % la
+        msg += "".join("\n\t\t- %s" % rule for rule in rules)
+
+        common_ancestors = self.__find_closest_common_ancestors(rules)
+        msg += "\n\nThe conflict may come from"
+
+        for rules, ancestors in common_ancestors.items():
+            a = ' or '.join("<%s>" % ancestor.name for ancestor in ancestors)
+            rule1, rule2 = sorted(rule.name for rule in rules)
+            msg += "\n\t\t- <%s> and <%s>: %s" % (rule1, rule2, a)
+
+        return GrammarError(msg)
+
+    def __find_closest_common_ancestors(self, conflicting_rules):
+        """
+        Find the closest common ancestor of pairs of rules.
+
+        :param conflicting_rules set: Set of Rule
+        :returns dict: 2-frozenset of NonTerminal rule names as keys, set of
+        NonTerminal ancestor names as values
+        """
+        parents = self.__rules_parents()
+        ancestors = {rule.origin: {rule.origin} for rule in conflicting_rules}
+        oldest_ancestors = {rule.origin: {rule.origin}
+                            for rule in conflicting_rules}
+        common_ancestors = dict()
+        intersections_found = set()
+
+        n = len(conflicting_rules)
+        pairs_number = (n * (n - 1)) // 2
+        while len(common_ancestors) < pairs_number:
```

Comment:
Is there a guarantee that we will always find that many pairs? Or could this turn into an infinite loop?

Reply:
1. The number of pairs in an n-set is `comb(n, 2) = n! / (2! * (n-2)!) = (n * (n-1)) / 2`, so `pairs_number` is exactly the number of pairs in `conflicting_rules`
2. Shift/Shift conflict arise if two rules could be matched in the same context. In other words, the conflicting rules are descendants of the starting rules in the pseudo-tree representing the relations between the rules. In particular, every pair taken from `conflicting_rules` has at least one common ancestor: the starting rule.
3. `ancestors` will eventually contain all the ancestors of each of the conflicting rules. Indeed, at the beginning of each step `i >= 0`, all of the ancestors of `rule` at a distance  `<= i` of `rule` are contained in `ancestors[rule.origin]`
4. `ancestors_intersection` builds the sets of common ancestors of rules given as arguments

These points prove that there are always exactly `pairs_number` pairs (tell me if one of them is not obvious). However, point 2 assumes that it is proven that Lark provides a valid LALR parser. For a non-complying parser (for example one which would find a conflict for rules that are not effectively used, and which could then be in different connected components of the rules graph), there is no guarantee on the termination of this loop.

If you want me to support non-complying parsers, we could brutally majorise the number of steps with `len(conflicting_rules)`. This has the drawback of hiding potential flaws in other parts of the parser.